README.md
[![checks](https://github.com/UCSF-Costello-Lab/LG3_Pipeline/actions/workflows/checks.yml/badge.svg)](https://github.com/UCSF-Costello-Lab/LG3_Pipeline/actions/workflows/checks.yml)

# LG3_Pipeline

_Warning: Because of hardcoded paths to annotation data files, and software dependencies on the [CBI software stack](https://www.c4.ucsf.edu/software/software-repositories.html), this software tool runs only on the UCSF [TIPCC] and [C4] compute clusters.  It is our long-term goal to make it run anywhere._


## What's new?

See [NEWS](NEWS.md) for the changlog.



## Instructions

The LG3 Pipeline is pre-installed on the [TIPCC] and the [C4] clusters.  To get access to it, load the following module:

```sh
$ module load CBI lg3
$ lg3 --version
2021-10-11
```

See `module avail` for alternative versions.


## Run test example

### Setup

To run through the built-in "small" test example (5-10 hours), let's create a separate project folder:
```sh
$ mkdir -p ~/lg3-demo
$ cd ~/lg3-demo
```

The first thing we want to do is to create an `output/` folder.  If we want it in the same location, we do:
```sh
$ mkdir output   ## folder where the output files will be saved
```
If we want it to be on a separate drive, we can create it there and then using a symbol link, e.g.
```sh
$ mkdir -p /another/drive/lg3-demo/output
$ ln -s /another/drive/lg3-demo/output output
```
In both cases, there will be a local `./output/` folder that the LG3 pipeline can write to.

The remaining parts of the test setup can be either be created automatically using `lg3 test setup` (the `lg3` command is in `${LG3_HOME}/bin`), or manually as one would do when analyzing other data than the test data.  To set up the test automatically, use:

```sh
$ export PATIENT=Patient157t10
$ lg3 test setup
Sourced: /path/to/LG3_Pipeline/lg3.conf
*** Setup
[OK] LG3_HOME: /path/to/LG3_Pipeline
[OK] R packages: 'RColorBrewer'
[OK] EMAIL: alice@example.org
[OK] PROJECT: LG3
[OK] PATIENT: Patient157t10 (required for 'lg3 test validate')
[OK] CONV (patient TSV file): patient_ID_conversions.tsv
[OK]   => SAMPLES: Z00599t10 Z00600t10 Z00601t10  (required by 'Recal' step)
[OK]   => NORMAL: 'Z00599t10' (required by 'Recal' step)
[OK] Raw data folder: rawdata
[OK] LG3_OUTPUT_ROOT: output
```

From the above, we should have a directory containing the following files and folders:
```sh
$ tree
.
|-- output
|-- patient_ID_conversions.tsv -> ~/lg3-demo/runs_demo/patient_ID_conversions.tsv
`-- rawdata -> /costellolab/data1/shared/LG3_Pipeline/example_data/rawdata

2 directories, 19 files
```


### Running the tests

**Importantly**, before starting, we need to set the `EMAIL` environment variables to an email address where job notifications are sent.
```sh
$ export EMAIL=alice@example.org
```
This can preferably be set in your global `~/.bashrc` script.


Now, we are ready to launch the pipeline (step by step):

``` sh
$ cd ~/lg3-demo
$ module load CBI lg3
$ export PATIENT=Patient157t10
$ lg3 run Trim                      ## ~5 minutes
$ lg3 run Align_gz                  ## ~5-10 minutes
$ lg3 run Recal                     ## ~2.5 hours
$ lg3 run Pindel && lg3 run MutDet  ## ~20 minutes & ~1.0 hour
$ lg3 run PostMut                   ## ~5 minutes
```

Optionally we can run the [exomeQualityPlots](https://github.com/SRHilz/exomeQualityPlots) pipeline developed by Stephanie Hilz (UCSF).
The pipeline generates quality plots of exome libraries and quality stats for mutation calling.
First we need to clone original exomeQualityPlots pipeline somewhere, e.g.
``` sh
$ mkdir -p ~/pipelines/exomeQualityPlots 
$ cd ~/pipelines/exomeQualityPlots 
$ git clone git@github.com:SRHilz/exomeQualityPlots.git
$ cd ${LG3_HOME}
$ ln -s ~/pipelines/exomeQualityPlots exomeQualityPlots
```

Now we are ready to roll:
``` sh
$ lg3 run QC_1 && lg3 run QC_2
$ lg3 run QC_3
```

Another option is to run [Costello-PSCN-Seq](https://github.com/HenrikBengtsson/Costello-PSCN-Seq) pipeline created by Henrik Bengtsson. The pipline implements Parent-specific copy number (PSCN) analysis on paired tumor-normal samples.
First we need to clone Costello-PSCN-Seq pipeline somewhere, e.g.
``` sh
$ mkdir -p ~/pipelines/Costello-PSCN-Seq
$ cd ~/pipelines/Costello-PSCN-Seq
$ git clone git@github.com:HenrikBengtsson/Costello-PSCN-Seq.git
$ cd ${LG3_HOME}
$ ln -s ~/pipelines/Costello-PSCN-Seq Costello-PSCN-Seq
```

Now we are ready to run:
``` sh
$ lg3 run PSCN
```

_Note_, all steps should be ran sequentially, except steps `Pindel` and `MutDet`, which can be run in parallel (as soon as the `Recal` step has finished).

_Tip:_ Each step of the pipeline is submitted to the Torque/PBS scheduler requesting a default number of cores (`nodes=1:ppn=...`) and amount of memory (`vmem=...`).  For now, you need to follow the source code to see what these defaults are.  You can override the defaults via environment variable `QSUB_OPTS`, e.g.
```sh
QSUB_OPTS="-l nodes=1:ppn=6 -l vmem=32gb" lg3 run Align_gz
```


### Checking progress and status

Throughout all steps, you can check the current status using the `lg3 status` command.  Here is what the output looks like when all steps are complete:
```sh
$ export PATIENT=Patient157t10
$ lg3 status
Sourced: /path/to/LG3_Pipeline/lg3.conf
Checking output for project LG3
Patient/samples table patient_ID_conversions.tsv
BAM suffix bwa.realigned.rmDups.recal.insert_size_metrics
Patients Patient157t10
****** Checking Patient157t10 Normal: Z00599t10
Fastq Z00599t10  OK
Fastq Z00600t10  OK
Fastq Z00601t10  OK
Trim Z00599t10  OK
Trim Z00600t10  OK
Trim Z00601t10  OK
BWA Z00599t10  OK
BWA Z00600t10  OK
BWA Z00601t10  OK
Recal Z00599t10  OK
Recal Z00600t10  OK
Recal Z00601t10  OK
UG  OK 7262
Germline Z00600t10  OK
Germline Z00601t10  OK
Pindel  OK 0
Mutect Z00600t10  OK 24
Mutect Z00601t10  OK 29
MutCombine  OK 6
MAF  OK
LOH plots  OK
```

Done!  All results are written to the `./output/` folder in different subdirectories.


See [Demo_output.md](run_demo/Demo_output.md) for a summary of what the output looks like.  Particularly, the text file `./output/LG3/MutInDel/Patient157t10.R.mutations` contain the identified mutations.



### Validate test results

To validate that you get the expected results when running through the tests, call `lg3 test validate`.  Here is an example of the output when all steps are completed:

```
$ export PATIENT=Patient157t10
$ lg3 test validate $PATIENT
Sourced: /path/to/LG3_Pipeline/lg3.conf
*** Configuration
[OK] PROJECT=LG3
[OK] PATIENT=Patient157-t10-underscore
[OK] CONV=patient_ID_conversions.tsv
[OK] LG3_TEST_TRUTH=/costellolab/data1/shared/LG3_Pipeline/example_data

*** Trimming of FASTQ Files
[OK] file tree ('output/LG3/trim/Z00*-trim')
[OK] file sizes ('output/LG3/trim/Z00*-trim/*')

*** BWA Alignment of FASTQ Files
[OK] file tree ('output/LG3/exomes')
[OK] file sizes ('output/LG3/exomes/Z00*/*')

*** Recalibration of BAM Files
[WARN] unexpected file tree ('/costellolab/data1/shared/LG3_Pipeline/example_data/Patient157-t10-underscore/output/LG3/exomes_recal/Patient157-t10-underscore' != 'output/LG3/exomes_recal/Patient157-t10-underscore')
@@ -2 +2,39 @@
-└── germline
+├── germline
+│   ├── NOR-Z00599_t10_AATCCGTC_L007_vs_Z00600_t10_AATCCGTC_L007.germline
+│   ├── NOR-Z00599_t10_AATCCGTC_L007_vs_Z00601_t10_AATCCGTC_L007.germline
+2.1M   output/LG3/exomes_recal/Patient157-t10-underscore/germline
+8.8K   output/LG3/exomes_recal/Patient157-t10-underscore/Z00599_t10_AATCCGTC_L007.bwa.realigned.rmDups.recal.insert_size_metrics
+1.8K   output/LG3/exomes_recal/Patient157-t10-underscore/Z00600_t10_AATCCGTC_L007.bwa.realigned.rmDups.recal.hybrid_selection_metrics
+12K    output/LG3/exomes_recal/Patient157-t10-underscore/Z00600_t10_AATCCGTC_L007.bwa.realigned.rmDups.recal.insert_size_histogram.pdf
+8.4K   output/LG3/exomes_recal/Patient157-t10-underscore/Z00600_t10_AATCCGTC_L007.bwa.realigned.rmDups.recal.insert_size_metrics

*** Recalibration of BAM Files
[OK] file tree ('output/LG3/exomes_recal/Patient157t10')
[OK] file sizes ('output/LG3/exomes_recal/Patient157t10/*')
[OK] file sizes ('output/LG3/exomes_recal/Patient157t10/germline/*')
[OK] file sizes ('output/LG3/exomes_recal/Patient157t10/*.bai')

*** Pindel Processing
[OK] file tree ('output/LG3/pindel')
[OK] file rows ('output/LG3/pindel/Patient157t10.pindel.cfg')
[OK] file sizes ('output/LG3/pindel/Patient157t10_pindel/*')

*** MutDet Processing
[OK] file tree ('output/LG3/mutations/Patient157t10_mutect')
[OK] file sizes ('output/LG3/mutations/Patient157t10_mutect/*')

*** Post-MutDet Processing
[OK] file tree ('output/LG3/MAF')
[OK] file sizes ('output/LG3/MAF/Patient157t10_MAF/*')
[OK] file sizes ('output/LG3/MAF/Patient157t10_plots/*')
[OK] file tree ('output/LG3/MutInDel')
[OK] file sizes ('output/LG3/MutInDel/*')
[OK] file content ('output/LG3/MutInDel/Patient157t10.R.mutations')
```

_Comment:_ There might minor discrepancies, which is due to these tests of file sizes sometimes being slightly to strict.  Regardless, if you get all `OK` for the content of `output/LG3/MutInDel/Patient157t10.R.mutations`, which contains the set of identified mutations, then you reproduced the expected results.



## Appendix

### Installation notes

The pipeline is installed on the TIPCC cluster, by cloning the [LG3_Pipeline] git repository and running the setup once, i.e.

```sh
$ cd /path/to/
$ git clone https://github.com/UCSF-Costello-Lab/LG3_Pipeline.git
$ cd LG3_Pipeline
$ make setup
```

The above folder is now where the LG3 Pipeline lives.  Environment variable `LG3_HOME` is set to point to this folder, e.g.

```sh
export LG3_HOME=/path/to/LG3_Pipeline
```

Then, in order to get access to the `lg3` command-line tool, make sure to also set:

```sh
export PATH="${LG3_HOME}/bin:${PATH}"
```


### Contributors

The following people (in reverse chronological order) have contributed to the LG3 Pipeline code base over the years:

* Henrik Bengtsson (2015-)
* Ivan Smirnov (2014-)
* Tali Mazor (2012-2017)
* Brett Johnson (2012-2014)
* Barry Taylor (2012-2013)
* Jun Song (2010-2011)


[LG3_Pipeline]: https://github.com/UCSF-Costello-Lab/LG3_Pipeline
[C4]: https://www.c4.ucsf.edu/
[TIPCC]: https://ucsf-ti.github.io/tipcc-web/


NEWS.md
# LG3_Pipeline

## Version 2021-10-11-9000 (develop version)

 * ...


## Version 2021-10-11

## BUG FIXES

 * `./lg3.conf` would not found shell scripts if the working
   directory was changed before calling the script.  This would
   happen with for instance `lg3 run PostMut`.

 * `lg3 run MutDet` broke in version 2021-10-10 because the internal
   `assert_*()` functions overwrote global variables. Now all
   internal Bash functions declare local variables explicitly.


## Version 2021-10-10

### SIGNFICANT CHANGES

 * Add `lg3 run <step>` command for launching the different LG3
   steps, e.g. `lg3 run Trim` and `lg3 run Align_gz`.

 * All software dependencies are now obtained from the CBI software
   stack available via `module load CBI`.

 * `lg3 test setup` no longer create symbolic links to run scripts,
   because we can now use `lg3 run <step>` instead.

### NEW FEATURES

 * Assert errors on software tools are now much more informative,
   e.g. the error message includes also the name of the environment
   variable that specifies the tool.

### BUG FIXES

 * User's ~/.Rprofile could wreak havoc during `lg3 run Recal` when
   calling Picard's CollectMultipleMetrics.jar, which calls `Rscript`.
   We're now using `R_PROFILE_USER=NULL` when calling this substep
   to emulate `Rscript --no-init-file`.


## Version 2021-10-09

### SIGNFICANT CHANGES

 * This is the first release that we managed to run on both TIPCC
   and C4.

### NEW FEATURES

 * Now the internal `LG3_SCRATCH_ROOT` path is agile to type of
   scheduler, i.e. Torque/PBS or Slurm.

 * Adding `-l walltime=hh:mm:ss` and `-l mem=<size>gb` declarations
   to PBS scripts to handle schedulers with hard resource limits.

 * Now `qsub` calls are agile to whether the underlying scheduler
   supports command-line option `-d <path>` or not.

### BUG FIXES

 * User's ~/.Rprofile could wreak havoc for some R scripts, e.g. by
   outputting messages.  Now all Rscript calls use '--vanilla' to make
   sure to run in an vanilla R session without user-specific settings.

 * Python script 'scripts/combine_snvs.py' used a hard-coded path
   to the 'Rscript' executable.

 * One Python script and one Rscript used a hard-coded path to
   bedtools executables.

 * lg3-test used a hardcoded path for the 'Rscript' executable.


## Version 2021-10-08

### NEW FEATURES

 * Now `pindel_all.pbs` reports also on the `ANNOVAR_HOME` software.

 * Now all `runs_demo/run_*` scripts pass also environment variables
   specifying software tools to the submitted job scripts.  This makes it
   possible to override the default software tools.

 * Now it's possible to override the _internal_ environment variables that
   the LG3 pipeline uses, e.g. `export MUTECT=/path/to/muTect-99.jar`.
   Note that this should only be used for troubleshooting and development.

### BUG FIXES

 * Internal assertions incorrectly required that the GATK and the MUTECT JAR
   files should be executable, but that is not required.


## Version 2020-05-26

### NEW FEATURES

 * `lg3 test validate` now compares the MD5 checksums of the content of the
   trimmed FASTQ files, the BWA aligned BAM files and the corresponding BAM
   index files. This requires that MD5 checksum files can be written.

### SOFTWARE QUALITY

 * The 'java', 'python' and 'Rscript' executables are now set in a central
   location to guarantee that the same, expected versions are used everywhere.

 * The AnnoVar, bedtools, cutadapt, BWA, GATK, MuTect, Picard, and Samtools
   executables are now set in a central location to guarantee that the same,
   expected versions are used everywhere.


## Version 2020-05-16

### NEW FEATURES

 * The LG3 Pipeline requires Python 2.  If an incompatible Python version is
   detected on the 'PATH', then an informative error is produced.
   
### KNOWN ISSUES

 * The 'lg3.conf' file must not be edited while the pipeline is running.  If
   done, then the outcome and the results are unpredictable.  This is because
   the 'lg3.conf' file is not frozen when the pipeline is launched.


## Version 2019-07-22

### SIGNIFICANT CHANGES

 * The LG3 Pipeline now refuses to run from within its installation folder,
   i.e. when the current working directory equals '${LG3_HOME}'.  This
   protects against various potential mistakes such as overriding installed
   files and settings.

 
### NEW FEATURES

 * Globals settings for the LG3 Pipeline such as locations of software tools
   are now configured in the '${LG3_HOME}/lg3.conf' bash script (which should
   not be edited by the user).  If a file 'lg3.conf' exists in the current
   working directory ("the project folder"), then that file is sourced after
   '${LG3_HOME}/lg3.conf', which makes it possible to override some or all
   of the predefined global settings on a project to project basis.  The
   latter file can also be used to configure variables such as PATIENT etc.

 * Now 'lg3 status' defaults to using '--all'.

 * ROBUSTNESS: Now `bin/lg3-test` explicitly asserts that Rscript exists
   before attemption to use it.


## Version 2019-03-23

### NEW FEATURES

 * Added option to use new Pindel 0.2.5b8 instead of Pindel 0.2.4t.

 * Added '_run_Align_mem' for an alternative data pre-processing compliant
   with Best Practices 2019.

 * Added '_run_Mutect2' for somatic mutation calling by Mutect2 and GATK 4.1.

 * Added '_run_Align_gz_no_trim' to align without trimming FASTQ data.

 * Added '_run_PSCN' for running the PSCN pipeline directly within the LG3
   pipeline.

### SOFTWARE QUALITY

 * Paths to resources and some other parameters are now in config file
   'lg3.conf' (work in progress).
 

## Version 2018-12-27

### NEW FEATURES

 * More informative and consistent error messages are provided in more place by
   making more use of internal utility functions such as 'error', 'warn',
   'assert_file_exists', 'assert_directory_exists', 'make_dir' and 'change_dir'.
  
### BUG FIXES

 * Recall_pass2.pbs failed to create symbolic link(s).



scripts/utils.sh
#!/bin/bash

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# OUTPUT
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Use colored stdout if the terminal supports it
## and as long as a stdout are not redirected
function term_colors {
    local action
    local what
    
    action=$1
    what=$2
    [[ -z "${what}" ]] && what=1
    
    if [[ "${action}" == "enable" && -t "${what}" ]]; then
        ## ANSI foreground colors
        black=$(tput setaf 0)
        red=$(tput setaf 1)
        green=$(tput setaf 2)
        yellow=$(tput setaf 3)
        blue=$(tput setaf 4)
        magenta=$(tput setaf 5)
        cyan=$(tput setaf 6)
        white=$(tput setaf 7)

        ## Text modes
        bold=$(tput bold)
        dim=$(tput dim)
        reset=$(tput sgr0)
    else
        export black=
        export red=
        export green=
        export yellow=
        export blue=
        export magenta=
        export cyan=
        export white=

        export bold=
        export dim=

        export reset=
    fi
}


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# TESTING
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
function test_context {
    local magenta
    local reset
    if [[ -t 1 ]]; then
        magenta=$(tput setaf 5)
        reset=$(tput sgr0)
    fi
    echo -e "${magenta}*** $*${reset}"
}

function test_context_begin {
    test_context "$* ..."
}

function test_context_end {
    test_context "$* ... done"
}


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# CONDITIONS
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
function error {
    local bold
    local gray
    local red
    local reset
    
    ON_ERROR=${ON_ERROR:-on_error}
    TRACEBACK_ON_ERROR=${TRACEBACK_ON_ERROR:-true}
    EXIT_ON_ERROR=${EXIT_ON_ERROR:-true}
    EXIT_VALUE=${EXIT_VALUE:-1}

    ## Parse arguments
    while [ -n "$1" ]; do
        case "$1" in
            --dryrun) EXIT_ON_ERROR=false; shift;;
            --value=*) EXIT_VALUE="${1/--value=/}"; shift;;
            *) break;;
        esac
    done

    if [[ -t 1 ]]; then
        red=$(tput setaf 1)
        gray=$(tput setaf 8)
        bold=$(tput bold)
        reset=$(tput sgr0)
    fi

    echo -e "${red}${bold}ERROR:${reset} ${bold}$*${reset}"

    if ${TRACEBACK_ON_ERROR}; then
       echo -e "${gray}Traceback:"
       for ((ii = 1; ii < "${#BASH_LINENO[@]}"; ii++ )); do
           printf "%d: %s() on line #%s in %s\\n" "$ii" "${FUNCNAME[$ii]}" "${BASH_LINENO[$((ii-1))]}" "${BASH_SOURCE[$ii]}"
       done
    fi

    if [[ -n "${ON_ERROR}" ]]; then
        if [[ $(type -t "${ON_ERROR}") == "function" ]]; then
            ${ON_ERROR}
        fi
    fi

    ## Exit?
    if ${EXIT_ON_ERROR}; then
        echo -e "Exiting (exit ${EXIT_VALUE})${reset}";
        exit "${EXIT_VALUE}"
    fi

    printf "%s" "${reset}"
}

function warn {  
    local bold
    local yellow
    local reset
    
    TRACEBACK_ON_WARN=${TRACEBACK_ON_WARN:-false}
    
    if [[ -t 1 ]]; then
        yellow=$(tput setaf 3)
        bold=$(tput bold)
        reset=$(tput sgr0)
    fi
    
    echo -e "${yellow}${bold}WARNING${reset}: $*"
    
    if ${TRACEBACK_ON_WARN}; then
       echo -e "${gray}Traceback:"
       for ((ii = 1; ii < "${#BASH_LINENO[@]}"; ii++ )); do
           printf "%d: %s() on line #%s in %s\\n" "$ii" "${FUNCNAME[$ii]}" "${BASH_LINENO[$((ii-1))]}" "${BASH_SOURCE[$ii]}"
       done
    fi
    
    printf "%s" "${reset}"
}


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# ASSERTIONS
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Usage: assert_file_exists /path/to/file
function assert_file_exists {
    local file
    local name
    local prefix
    
    [[ $# -ne 1 ]] && error "${FUNCNAME[0]}() requires a single argument: $#"
    file=$1
    prefix=""
    if [[ $file == --by-env=* ]]; then
	name=${file##--by-env=}
	file=${!name}
	prefix="$name="
    fi
    [[ -n "$file" ]] || error "File name must be non-empty: $prefix'$file'"
    [[ -f "$file" ]] || error "No such file: $prefix'$file' (working directory '${PWD}')"
}

function assert_link_exists {
    local file
    local name
    local prefix
    
    [[ $# -ne 1 ]] && error "${FUNCNAME[0]}() requires a single argument: $#"
    [[ -n "$1" ]] || error "File name must be non-empty: '$1'"
    file=$1
    prefix=""
    if [[ $file == --by-env=* ]]; then
	name=${file##--by-env=}
	file=${!name}
	prefix="$name="
    fi
    [[ -L "$file" ]] || error "File is not a link: $prefix'$file' (working directory '${PWD}')"
    [[ -e "$file" ]] || error "[File] link is broken: $prefix'$file' (working directory '${PWD}')"
}

## Usage: assert_file_executable /path/to/file
function assert_file_executable {
    local file
    local name
    local prefix
    
    [[ $# -ne 1 ]] && error "${FUNCNAME[0]}() requires a single argument: $#"
    assert_file_exists "$1"
    file=$1
    prefix=""
    if [[ $file == --by-env=* ]]; then
	name=${file##--by-env=}
	file=${!name}
	prefix="$name="
    fi
    [[ -x "$file" ]] || error "File exists but is not executable: $prefix'$file' (working directory '${PWD}')"
}

## Usage: assert_directory_exists /path/to/folder
function assert_directory_exists {
    local dir
    local name
    local prefix
    
    [[ $# -ne 1 ]] && error "${FUNCNAME[0]}() requires a single argument: $#"
    dir=$1
    prefix=""
    if [[ $dir == --by-env=* ]]; then
	name=${dir##--by-env=}
	dir=${!name}
	prefix="$name="
    fi
    [[ -n "$dir" ]] || error "Directory name must be non-empty: $prefix='$dir'"
    [[ -d "$dir" ]] || error "No such directory: $prefix='$dir' (working directory '${PWD}')"
}

## Usage: assert_patient_name "${PATIENT}"
function assert_patient_name {
    [[ $# -ne 1 ]] && error "${FUNCNAME[0]}() requires a single argument: $#"
    [[ -n "$1" ]] || error "Patient name must be non-empty: '$1'"
    [[ "$1" == *[_]* ]] && error "Patient name must not contain underscores: $1"
}


## Usage: assert_pwd
function assert_pwd {
    ## Need to disable this assertion to avoid triggering a false SC2119 ShellCheck warning
    ## [[ $# -ne 0 ]] && error "${FUNCNAME[0]}() must not be called with arguments: $#"
    
    ## Don't allow running the pipeline from within LG3_HOME
    equal_dirs "${PWD}" "${LG3_HOME}" && error "The LG3 Pipeline must not be run from the folder where it is installed (LG3_HOME): ${PWD}"
}


## Usage: assert_python "" or assert_python "<python-binary>"
function assert_python {
    local bin
    local name
    local prefix
    local version
    local version_x_y
    
    ## Arguments are optional
    bin=$1

    prefix=""
    if [[ $bin == --by-env=* ]]; then
	name=${bin##--by-env=}
	bin=${!name}
	prefix="$name="
    fi
    
    
    if [[ -n "$bin" ]]; then
        assert_file_executable "$1"
    else
        bin=$(command -v python) || error "Python executable not found on PATH: ${PATH}"
    fi
    
    ## Assert correct version
    version=$(2>&1 "$bin" --version | sed -E 's/.*(P|p)ython *//g')
    version_x_y=$(echo "$version" | sed -E 's/[.][0-9]+$//g')
    [[ "$version_x_y" == "2.6" ]] || [[ "$version_x_y" == "2.7" ]] || error "Requires Python 2.6 or 2.7: $version ($bin)"
}

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# NAVIGATION
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
function change_dir {
    local opwd
    
    opwd=${PWD}
    assert_directory_exists "$1"
    cd "$1" || error "Failed to set working directory to $1"
    echo "New working directory: '$1' (was '${opwd}')"
}

function make_dir {
    mkdir -p "$1" || error "Failed to create new working directory $1"
}

function make_change_dir {
    make_dir "$1"
    change_dir "$1"
}

function equal_dirs {
    local a
    local b
    a=$(readlink -f "$1")
    b=$(readlink -f "$2")
    [[ "${a}" == "${b}" ]]
}


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# LG3 specific
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
function source_lg3_conf {
    ## Already sourced?
    if [[ -n ${LG3_CONF_SOURCED} ]]; then
	echo "All lg3.conf files have already been sourced and validates. Skipping."
	return 0;
    fi
	
    ## The default settings
    assert_file_exists "${LG3_HOME}/lg3.conf"
    # shellcheck disable=1090
    source "${LG3_HOME}/lg3.conf"
    echo "Sourced: ${LG3_HOME}/lg3.conf"

    ## Settings specific to the project folder?
    if [ -f "lg3.conf" ] && ! equal_dirs "." "${LG3_HOME}"; then
        # shellcheck disable=1090
        source "lg3.conf"
        echo "Sourced: ${PWD}/lg3.conf ($(stat --printf='%s' lg3.conf) bytes)"
    fi

    ## Assert that all LG3 software dependencies are valid
    lg3_assert_software
    
    LG3_CONF_SOURCED=true
}



function lg3_software_envvars {
    echo "JAVA PYTHON RSCRIPT ANNOVAR_HOME BEDTOOLS BWA CUTADAPT GATK MUTECT PICARD_HOME SAMTOOLS PINDEL PINDEL2VCF TG"
}

function lg3_list_software {
    local name
    local value
    
    echo "LG3 software dependencies used:"
    for name in $(lg3_software_envvars); do
        value=${!name}
	if [[ -z "${value}" ]]; then
            value="<empty>"
	else
            value="'${value}'"
	fi
        echo "- ${name}=${value}"
    done
}

function lg3_assert_software {
    assert_file_executable  --by-env=JAVA
    assert_file_executable  --by-env=PYTHON
    assert_file_executable  --by-env=PYTHON
    assert_python           --by-env=PYTHON
    assert_file_executable  --by-env=RSCRIPT
    assert_directory_exists --by-env=ANNOVAR_HOME
    assert_file_executable  --by-env=BEDTOOLS
    assert_file_executable  --by-env=BWA
    assert_file_executable  --by-env=CUTADAPT
    assert_file_exists      --by-env=GATK
    assert_file_exists      --by-env=MUTECT
    assert_directory_exists --by-env=PICARD_HOME
    assert_file_executable  --by-env=SAMTOOLS
    assert_file_executable  --by-env=PINDEL
    assert_file_executable  --by-env=PINDEL2VCF
    assert_file_executable  --by-env=TG
}

function lg3_qsub_envvar_append_software {
    local name
    local value
    local res

    lg3_assert_software
    
    res="${QSUB_ENVVARS}"
    for name in $(lg3_software_envvars); do
        value=${!name}
	if [[ -n "${value}" ]]; then
            res="${res},${name}=${value}"
	fi
    done
    
    echo "${res}"
}


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# Job scheduler specifics
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Does 'qsub' on the current system support CLI option -d <path>?
qsub_can_set_pwd=false
if [[ ${CLUSTER} == "tipcc" ]]; then
    # shellcheck disable=2034
    qsub_can_set_pwd=true
fi


if [[ -n "${PBS_JOBID}" ]]; then
    LG3_SCRATCH_ROOT=${LG3_SCRATCH_ROOT:-/scratch/${USER:?}/${PBS_JOBID}}
elif [[ -n "${SLURM_JOB_ID}" ]]; then
    LG3_SCRATCH_ROOT=${LG3_SCRATCH_ROOT:-/scratch/${USER:?}/${SLURM_JOB_ID}}
elif [[ -n "${TMPDIR}" ]]; then
    LG3_SCRATCH_ROOT=${LG3_SCRATCH_ROOT:-${TMPDIR}}
else
    error "Cannot set LG3_SCRATCH_ROOT"
fi


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# MAIN
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## FIXME: Can we remove this?
unset PYTHONPATH  ### ADHOC: In case it is set by user. /HB 2018-09-07
